From 539fd85c4236495b1a68fc0fbf3040b82982d3e4 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Wed, 15 Aug 2018 07:52:46 +0200
Subject: [PATCH 1/1] efi_loader: memory leak in efi_set_bootdev()

efi_set_bootdev() may be called repeatedly.
Free the memory allocated for device paths in previous calls.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 cmd/bootefi.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/cmd/bootefi.c b/cmd/bootefi.c
index c5a2bd970d..1dffb8ab11 100644
--- a/cmd/bootefi.c
+++ b/cmd/bootefi.c
@@ -582,6 +582,13 @@ void efi_set_bootdev(const char *dev, const char *devnr, const char *path)
 	char filename[32] = { 0 }; /* dp->str is u16[32] long */
 	char *s;
 
+	/* efi_set_bootdev is typically called repeatedly, recover memory */
+	efi_free_pool(bootefi_device_path);
+	efi_free_pool(bootefi_image_path);
+	/* If blk_get_device_part_str fails, avoid duplicate free. */
+	bootefi_device_path = NULL;
+	bootefi_image_path = NULL;
+
 	if (strcmp(dev, "Net")) {
 		struct blk_desc *desc;
 		disk_partition_t fs_partition;
-- 
2.18.0

